oci_hook.go

```diff
@@ -0,0 +1,126 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// OciHook represents an OCI hook, including its required parameters.
+type OciHook struct {
+	Path string
+	Args []string
+}
+
+// OciHooks gathers all existing OCI hooks list.
+type OciHooks struct {
+	PreStartHooks  []OciHook
+	PostStartHooks []OciHook
+	PostStopHooks  []OciHook
+}
+
+func buildOciHookState(processID int) specs.State {
+	return specs.State{
+		Pid: processID,
+	}
+}
+
+func runOciHook(hook OciHook) error {
+	firstArg := hook.Args[0]
+	otherArgs := hook.Args[1:]
+	args := []string{"-a", firstArg, hook.Path}
+	args = append(args, otherArgs...)
+
+	cmd := exec.Command("exec", args...)
+
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	state := buildOciHookState(os.Getpid())
+	stateJSON, err := json.Marshal(state)
+	if err != nil {
+		return err
+	}
+
+	err = cmd.Start()
+	if err != nil {
+		return err
+	}
+
+	_, err = stdin.Write(stateJSON)
+	if err != nil {
+		return err
+	}
+
+	err = cmd.Wait()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *OciHooks) preStartHooks() error {
+	if len(h.PreStartHooks) < 1 {
+		return nil
+	}
+
+	for _, hook := range h.PreStartHooks {
+		err := runOciHook(hook)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *OciHooks) postStartHooks() error {
+	if len(h.PostStartHooks) < 1 {
+		return nil
+	}
+
+	for _, hook := range h.PostStartHooks {
+		err := runOciHook(hook)
+		if err != nil {
```

Comment:
With posts start hooks, we only need to log the error and continue running the remaining hooks.